File: writejson.c
#include "gitin.h"

#include <git2/commit.h>
#include <stdio.h>

static void writesignature(FILE* fp, const git_signature* sig) {
	fprintf(fp, "{");
	hprintf(fp, "\"name\":\"%j\",", sig->name);
	hprintf(fp, "\"email\":\"%j\",", sig->email);
	hprintf(fp, "\"date\":\"%T\"", &sig->when);
	fprintf(fp, "}");
}

void writejsoncommit(FILE* fp, git_commit* commit, int first) {
	char                 oid[GIT_OID_SHA1_HEXSIZE + 1], parentoid[GIT_OID_SHA1_HEXSIZE + 1];
	const git_signature* author    = git_commit_author(commit);
	const git_signature* committer = git_commit_committer(commit);
	const char*          summary   = git_commit_summary(commit);
	const char*          message   = git_commit_message(commit);

	git_oid_tostr(oid, sizeof(oid), git_commit_id(commit));
	git_oid_tostr(parentoid, sizeof(parentoid), git_commit_parent_id(commit, 0));

	if (!first)
		fprintf(fp, ",");
	fprintf(fp, "\"%s\":", oid);

	fprintf(fp, "{");
	fprintf(fp, "\"id\":\"%s\",", oid);
	if (*parentoid)
		fprintf(fp, "\"parent\":\"%s\",", parentoid);
	else
		fprintf(fp, "\"parent\":null,");
	if (author) {
		fprintf(fp, "\"author\":");
		writesignature(fp, author);
		fprintf(fp, ",");
	} else {
		fprintf(fp, "\"author\":null,");
	}
	if (committer) {
		fprintf(fp, "\"committer\":");
		writesignature(fp, committer);
		fprintf(fp, ",");
	} else {
		fprintf(fp, "\"committer\":null,");
	}
	if (summary)
		hprintf(fp, "\"summary\":\"%j\",", summary);
	else
		fprintf(fp, "\"summary\":null,");
	if (message)
		hprintf(fp, "\"message\":\"%j\"", message);
	else
		fprintf(fp, "\"message\":null");
	fprintf(fp, "}");
}

void writejsonref(FILE* fp, const struct repoinfo* info, git_reference* ref, git_commit* commit) {
	int ishead = info->head && !git_oid_cmp(git_reference_target(ref), info->head);

	fprintf(fp, "{");
	hprintf(fp, "\"name\":\"%j\",", git_reference_shorthand(ref));
	hprintf(fp, "\"head\":%j,", ishead ? "true" : "false");
	fprintf(fp, "\"commit\":");
	writejsoncommit(fp, commit, 0);
	fprintf(fp, "}");
}

File: writejson.go
package gitin

import (
	"encoding/json"
	"io"
	"time"

	git "github.com/jeffwelling/git2go/v37"
)

type JSONSignature struct {
	Name  string    `json:"name"`
	Email string    `json:"email"`
	Date  time.Time `json:"date"`
}

type JSONCommit struct {
	ID        string         `json:"id"`
	Parent    *string        `json:"parent"`
	Author    *JSONSignature `json:"author"`
	Committer *JSONSignature `json:"committer"`
	Summary   *string        `json:"summary"`
	Message   *string        `json:"message"`
}

type JSONRef struct {
	Name   string     `json:"name"`
	Commit JSONCommit `json:"commit"`
}

type JSONOut struct {
	Branches []JSONRef    `json:"branches,omitempty"`
	Tags     []JSONRef    `json:"tags,omitempty"`
	Commits  []JSONCommit `json:"commits,omitempty"`
}

func toJSONSignature(sig *git.Signature) *JSONSignature {
	if sig == nil {
		return nil
	}
	return &JSONSignature{
		Name:  sig.Name,
		Email: sig.Email,
		Date:  sig.When,
	}
}

func strPtrNonEmpty(s string) *string {
	if s == "" {
		return nil
	}
	return &s
}

func toJSONCommit(c *git.Commit) JSONCommit {
	id := c.Id().String()

	var parent *string
	if c.ParentCount() > 0 {
		p := c.ParentId(0).String()

		if p != "" {
			parent = &p
		}
	}

	return JSONCommit{
		ID:        id,
		Parent:    parent,
		Author:    toJSONSignature(c.Author()),
		Committer: toJSONSignature(c.Committer()),
		Summary:   strPtrNonEmpty(c.Summary()),
		Message:   strPtrNonEmpty(c.Message()),
	}
}

func writejsonrefs(w io.Writer, info *repoinfo) error {
	out := JSONOut{
		Branches: make([]JSONRef, 0, len(info.refs)),
		Tags:     make([]JSONRef, 0),
	}
	for _, r := range info.refs {
		if r == nil || r.commit == nil {
			continue
		}
		jr := JSONRef{
			Name:   r.refname,
			Commit: toJSONCommit(r.commit),
		}
		if r.istag {
			out.Tags = append(out.Tags, jr)
		} else {
			out.Branches = append(out.Branches, jr)
		}
	}

	enc := json.NewEncoder(w)
	enc.SetIndent("", "  ")
	return enc.Encode(out)
}

func writejsoncommits(w io.Writer, c []JSONCommit) error {
	out := JSONOut{
		Commits: c,
	}
	enc := json.NewEncoder(w)
	enc.SetIndent("", "  ")
	return enc.Encode(out)
}
